Adds builders page

ryanviana opened the pull request:
#3 

## Description

Creates a builders page, using ethereum blockies to create a profile image to each builder. 

## Demo

https://youtu.be/N7WpVfpW4k8

## Additional Information

- [x] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [x] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)

Your ENS/address: 0xF3b64dD5AF39d8fF0c614F7637e339e31466c4C3

BillSchumacher commented on packages/nextjs/app/builders/page.tsx:
```diff
@@ -0,0 +1,80 @@
+
+        try {
+          const filter = contract.filters.CheckedIn(null, null, null);
+          const events = await contract.queryFilter(filter, -100000);
```
Event's probably a good way to handle it but maybe we should do it the scaffold eth way?

Like the dice game challenge:
```
  const { data: rollsHistoryData, isLoading: rollsHistoryLoading } = useScaffoldEventHistory({
    contractName: "DiceGame",
    eventName: "Roll",
    fromBlock: 0n,
  });
  ```

carletex replied:
yes! This way you don't have to worry about contract instanciation, keep track of the ABI & address... It's all handled for you.

ryanviana replied:
Thanks for the reviewing & insights, @BillSchumacher and @carletex! Submitted a new commit ;)

carletex reviewed (changes requested):
See comments, thanks!

technophile-04 commented on packages/nextjs/app/builders/page.tsx:
```diff
@@ -0,0 +1,73 @@
+            <div key={index} className="bg-white border-2 border-gray-200 shadow-sm rounded-lg overflow-hidden">
+              <div className="p-5">
+                <div className="mx-auto h-32 w-32 rounded-full shadow-lg mb-4 relative">
+                  <Image
+                    src={getBlockieImageUrl(builder)}
+                    alt={`Builder ${builder}`}
+                    layout="fill"
+                    objectFit="cover"
+                    className="rounded-full"
+                  />
```
We could use [`BlockieAvatar`](https://github.com/BuidlGuidl/batch4.buidlguidl.com/blob/main/packages/nextjs/components/scaffold-eth/BlockieAvatar.tsx) component directly  

Example usage [here](https://github.com/BuidlGuidl/batch4.buidlguidl.com/blob/main/packages/nextjs/components/scaffold-eth/Address.tsx#L91-L95)

If you want more control over styling you could directly use `blo` (already pre-installed in SE-2) and internally used by BlockieAvatar component to get img URL,  example [here](https://github.com/BuidlGuidl/batch4.buidlguidl.com/blob/ba6c637316c4c068a827afebe51c0fcade19f0b2/packages/nextjs/components/scaffold-eth/BlockieAvatar.tsx#L12)

technophile-04 commented on packages/nextjs/package.json:
```diff
@@ -21,6 +21,8 @@
     "@uniswap/v2-sdk": "^3.0.1",
     "blo": "^1.0.1",
     "daisyui": "4.5.0",
+    "ethereum-blockies-base64": "^1.0.2",
+    "ethers": "v5",
```
Lets remove both of this library, we already have good alternatives of this in SE-2 🙌

checkout https://github.com/BuidlGuidl/batch4.buidlguidl.com/pull/23/files#r1560902439

technophile-04 commented on packages/nextjs/app/builders/page.tsx:
```diff
@@ -0,0 +1,73 @@
+
+  useEffect(() => {
+    if (!isLoading && checkedInEvents) {
+      const builderAddresses = checkedInEvents
+        .map(event => event.args.builder as string | undefined)
+        .filter((builder): builder is string => !!builder);
+      const uniqueBuilders = Array.from(new Set(builderAddresses));
```
You don't need `useEffect` and an extra `builders`  state. 

You could move this out of `useEffect` and use `uniqueBuilders` directly while rendering on UI

technophile-04 commented on packages/nextjs/app/builders/page.tsx:
```diff
@@ -0,0 +1,73 @@
+      <div className="max-w-7xl mx-auto p-8 bg-white shadow-2xl rounded-2xl my-10">
+        <h1 className="text-4xl font-extrabold text-center text-gray-900 mb-10">Our Builders</h1>
+        <div className="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-4 gap-6">
+          {builders.map((builder, index) => (
```
Maybe show some loading indicator, here 

```tsx
{isLoading ? (
  <span className="loading loading-spinner"></span>
) : (
  uniqueBuilders?.map((builder, index) => (...) }
```

You could use daisyUI loading classes for this https://daisyui.com/components/loading/

technophile-04 commented:
Looking really nice @ryanviana ! Added a couple of comments 

Also maybe we should also add link to `/builders` page in Header component  beside Home , Debug Contracts nav links

BillSchumacher reviewed (approved):
Nice, [technophile-04](https://github.com/technophile-04) has some good suggestions too.

ryanviana commented:
Thanks for the comments & review, @technophile-04. I submitted some new commits addressing those. That was valuable insight for me. I'm new to scaffold & blockchain dev, this is being important for me to understand the real value of SE2.

ryanviana commented:
Just updated the demo video ;)

carletex reviewed (changes requested):
This is looking amazing @ryanviana !!

A couple of suggestions:

1. Some builders have ENS addresses... if they do, we should show that instead of their address. Check our `Address` component and the use of the ENS hooks
2. I think it'd more legible if you extract the code for
```tsx
<div className="p-5">
    <div className="mx-auto h-32 w-32 rounded-full shadow-lg mb-4 relative">
      <BlockieAvatar address={builder} size={128} />
    </div>
    <h3 className="text-lg text-black font-semibold text-center">Builder {index + 1}</h3>
    <p className="text-gray-600 text-sm text-center mt-1">Address: {truncateAddress(builder)}</p>
  </div>
  <div className="bg-gray-50 p-4">
    <Link href={`/builders/${builder}`} passHref>
      <span className="flex items-center justify-center gap-2 text-blue-600 hover:text-blue-800 transition-colors cursor-pointer">
        <GlobeAltIcon className="h-6 w-6" />
        View Profile
      </span>
    </Link>
  </div>
```

to a separate component. Something like `Batch4AddressDisplay` or similar. Abstract all the display & logic there (ENS, etc) and use it like `<Batch4AddressDisplay address={builder} />`

ryanviana commented:
> This is looking amazing @ryanviana !!
> 
> A couple of suggestions:
> 
> 1. Some builders have ENS addresses... if they do, we should show that instead of their address. Check our `Address` component and the use of the ENS hooks
> 2. I think it'd more legible if you extract the code for
> 
> ```tsx
> <div className="p-5">
>     <div className="mx-auto h-32 w-32 rounded-full shadow-lg mb-4 relative">
>       <BlockieAvatar address={builder} size={128} />
>     </div>
>     <h3 className="text-lg text-black font-semibold text-center">Builder {index + 1}</h3>
>     <p className="text-gray-600 text-sm text-center mt-1">Address: {truncateAddress(builder)}</p>
>   </div>
>   <div className="bg-gray-50 p-4">
>     <Link href={`/builders/${builder}`} passHref>
>       <span className="flex items-center justify-center gap-2 text-blue-600 hover:text-blue-800 transition-colors cursor-pointer">
>         <GlobeAltIcon className="h-6 w-6" />
>         View Profile
>       </span>
>     </Link>
>   </div>
> ```
> 
> to a separate component. Something like `Batch4AddressDisplay` or similar. Abstract all the display & logic there (ENS, etc) and use it like `<Batch4AddressDisplay address={builder} />`

Great feedback, Carlos. I submitted new commits addressing these issues. Also updated the demo video, but this time I had to upload it to YouTube, because it got a bit too heavy for Github. 

I also tried to implement night mode! Not that good with colors choice though. Hope you like it.

perstarkse commented:
Looking great @ryanviana!

I'm getting a console error and a warning when running it locally. 

```
Error: Contract not found   useScaffoldEventHistory.ts
``` 

Then it loads ok. Looking at the code examples from the [Scaffold-ETH 2 Docs](https://docs.scaffoldeth.io/hooks/useScaffoldEventHistory) I see no obvious error or solution.

The warning:
```
Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>. 
```

ryanviana commented:
> Looking great @ryanviana!
> 
> I'm getting a console error and a warning when running it locally.
> 
> ```
> Error: Contract not found   useScaffoldEventHistory.ts
> ```
> 
> Then it loads ok. Looking at the code examples from the [Scaffold-ETH 2 Docs](https://docs.scaffoldeth.io/hooks/useScaffoldEventHistory) I see no obvious error or solution.
> 
> The warning:
> 
> ```
> Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>. 
> ```

Strange man, I'm not being able to reproduce this error. I tried cloning a fresh repo in a new directory but it ran without problems.

perstarkse commented:
> > Looking great @ryanviana!
> > I'm getting a console error and a warning when running it locally.
> > ```
> > Error: Contract not found   useScaffoldEventHistory.ts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then it loads ok. Looking at the code examples from the [Scaffold-ETH 2 Docs](https://docs.scaffoldeth.io/hooks/useScaffoldEventHistory) I see no obvious error or solution.
> > The warning:
> > ```
> > Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>. 
> > ```
> 
> Strange man, I'm not being able to reproduce this error. I tried cloning a fresh repo in a new directory but it ran without problems.

Alright, thanks for checking. Will try and see if I can reproduce it for clarity.

carletex commented on packages/nextjs/app/builders/_components/Batch4Avatar.tsx:
```diff
@@ -0,0 +1,37 @@
+  }, [fetchedEnsAvatar]);
+
+  return (
+    <img
```
This is giving this linting error. You can ignore it with: // eslint-disable-next-line @next/next/no-img-element

![image](https://github.com/BuidlGuidl/batch4.buidlguidl.com/assets/2486142/178ade2a-de2b-48ac-80f7-ce617c7994d8)

carletex commented on packages/nextjs/app/builders/_components/Batch4AddressDisplay.tsx:
```diff
@@ -0,0 +1,36 @@
+  const isDarkMode = resolvedTheme === "dark";
+
+  return (
+    <div className={`flex flex-col items-center justify-center rounded-lg p-5 bg-white shadow-md`}>
```
You don't need the `{` in for the classNames, since you are not doing anything dynamic.

It can be just `className="..."`

carletex replied:
You can also "fix it" in other places you have the same

carletex commented on packages/nextjs/app/builders/_components/Batch4AddressDisplay.tsx:
```diff
@@ -0,0 +1,36 @@
+        <Batch4Avatar address={address} size={128} />
+      </div>
+      <h3 className={"text-lg font-semibold text-center text-black"}>Builder {index + 1}</h3>
+      <p className={"text-gray-600 text-sm text-center mt-1 text-black"}>
```
text-gray-600 and text-black override each other :D

carletex commented on packages/nextjs/app/builders/_components/Batch4AddressDisplay.tsx:
```diff
@@ -0,0 +1,36 @@
+        <Batch4Avatar address={address} size={128} />
+      </div>
+      <h3 className={"text-lg font-semibold text-center text-black"}>Builder {index + 1}</h3>
+      <p className={"text-gray-600 text-sm text-center mt-1 text-black"}>
+        <Batch4Address address={address} format="short" size="base" />
+      </p>
```
I believe this is causing this error (Should be a div wrapper, not a p)

![image](https://github.com/BuidlGuidl/batch4.buidlguidl.com/assets/2486142/ac92a34d-13dc-47d1-a0d3-5e1f857fa7af)

carletex reviewed (changes requested):
We are almost there @ryanviana good job!! Left a few comments.

Regarding the `Error: Contract not found   useScaffoldEventHistory.ts` error that @perstarkse was pointing, I think that's something on our end (Scaffold-ETH-2)... so not worry about it (since it's still working, just throwing an initial error when loading)

carletex commented on packages/nextjs/app/builders/page.tsx:
```diff
@@ -0,0 +1,56 @@
+    contractName: "BatchRegistry",
+    eventName: "CheckedIn",
+    fromBlock: 118484927n,
+    watch: true,
```
Also, I think we can remove this. It's going to cause a lot of RPC calls... and we don't need to be that dynamic in this case :D

ryanviana commented:
Thanks for the feedback, @carletex. Tried to address the points in the last commits! ;)

carletex commented:
Great stuff @ryanviana !! Merging this
